File: src/main/java/in/tazj/k8s/letsencrypt/acme/Route53Responder.java
package in.tazj.k8s.letsencrypt.acme;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.ImmutableList;

import com.amazonaws.services.route53.AmazonRoute53;
import com.amazonaws.services.route53.model.Change;
import com.amazonaws.services.route53.model.ChangeAction;
import com.amazonaws.services.route53.model.ChangeBatch;
import com.amazonaws.services.route53.model.ChangeResourceRecordSetsRequest;
import com.amazonaws.services.route53.model.GetChangeRequest;
import com.amazonaws.services.route53.model.HostedZone;
import com.amazonaws.services.route53.model.RRType;
import com.amazonaws.services.route53.model.ResourceRecord;
import com.amazonaws.services.route53.model.ResourceRecordSet;

import java.util.Optional;

import in.tazj.k8s.letsencrypt.util.LetsencryptException;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.Synchronized;
import lombok.experimental.var;
import lombok.extern.slf4j.Slf4j;
import lombok.val;

import static com.amazonaws.services.route53.model.ChangeAction.DELETE;
import static com.amazonaws.services.route53.model.ChangeAction.UPSERT;

/**
 * An ACME challenge responder using AWS Route53.
 */
@Slf4j
@RequiredArgsConstructor
public class Route53Responder implements DnsResponder {
  final private AmazonRoute53 route53;

  @Override
  public String addChallengeRecord(String recordName, String challengeDigest) {
    return updateRoute53Record(recordName, challengeDigest, UPSERT).getName();
  }

  @Override
  public void removeChallengeRecord(String recordName, String challengeDigest) {
    updateRoute53Record(recordName, challengeDigest, DELETE);
  }

  @Synchronized
  @SneakyThrows // Ignore InterruptedException
  private HostedZone updateRoute53Record(String recordName,
                                         String challengeDigest,
                                         ChangeAction action) {
    val hostedZone = findHostedZone(recordName);

    if (!hostedZone.isPresent()) {
      log.error("No hosted zone found for record: {}!", recordName);
      throw new LetsencryptException("No hosted zone found");
    }

    val recordValue = formatChallengeValue(challengeDigest);
    val recordSet = new ResourceRecordSet(recordName, RRType.TXT)
        .withTTL(60L)
        .withResourceRecords(new ResourceRecord(recordValue));
    val changeBatch = new ChangeBatch(ImmutableList.of(
        new Change(action, recordSet)));

    val request =
        new ChangeResourceRecordSetsRequest(hostedZone.get().getId(), changeBatch);

    /* Commit change and wait until AWS confirms propagation */
    var changeInfo =
        route53.changeResourceRecordSets(request).getChangeInfo();

    log.info("Applied {} to record {} in zone {} with value {}. Status: {}",
        action, recordName, hostedZone.get().getName(), recordValue, changeInfo.getStatus());

    while (changeInfo.getStatus().equals("PENDING")) {
      changeInfo = route53.getChange(new GetChangeRequest(changeInfo.getId())).getChangeInfo();
      Thread.sleep(1000);
    }

    return hostedZone.get();
  }

  /**
   * If no hosted zone is supplied explicitly this function will attempt to find a managed zone in
   * AWS that matches the supplied record.
   * If multiple matching zones are found the most specific one is chosen.
   */
  @VisibleForTesting
  public Optional<HostedZone> findHostedZone(String record) {
    val fqdnRecord = determineFqdnRecord(record);
    // Attempt to find the correct hosted zone by matching the longest matching suffix.
    val matchingZone = route53.listHostedZones().getHostedZones().stream()
        .filter(zone -> fqdnRecord.endsWith(zone.getName()))
        .reduce(((acc, zone) -> {
          if (zone.getName().length() > acc.getName().length()) {
            return zone;
          }
          return acc;
        }));

    matchingZone.ifPresent(zone ->
        log.info("Found matching zone {} for record {}", zone.getName(), record));

    return matchingZone;
  }

  // Route53 requires TXT record values to be explicitly enclosed in quotation marks.
  private static String formatChallengeValue(String challenge) {
    if (challenge.endsWith("\"")) {
      return challenge;
    } else {
      return "\"" + challenge + "\"";
    }
  }
}

File: src/main/kotlin/in/tazj/k8s/letsencrypt/acme/Route53Responder.kt
package `in`.tazj.k8s.letsencrypt.acme

import `in`.tazj.k8s.letsencrypt.util.LetsencryptException
import com.amazonaws.services.route53.AmazonRoute53
import com.amazonaws.services.route53.model.*
import com.amazonaws.services.route53.model.ChangeAction.DELETE
import com.amazonaws.services.route53.model.ChangeAction.UPSERT
import com.google.common.collect.ImmutableList
import org.funktionale.option.Option
import org.funktionale.option.toOption
import org.slf4j.LoggerFactory


/**
 * An ACME challenge responder using AWS Route53.
 */
class Route53Responder(val route53: AmazonRoute53) : DnsResponder {
    val log = LoggerFactory.getLogger(this::class.java)

    override fun addChallengeRecord(recordName: String, challengeDigest: String): String {
        return updateRoute53Record(recordName, challengeDigest, UPSERT).name
    }

    override fun removeChallengeRecord(recordName: String, challengeDigest: String) {
        updateRoute53Record(recordName, challengeDigest, DELETE)
    }

    @Synchronized
    private fun updateRoute53Record(recordName: String,
                                    challengeDigest: String,
                                    action: ChangeAction): HostedZone {
        val hostedZone = findHostedZone(recordName)

        if (hostedZone.isEmpty()) {
            log.error("No hosted zone found for record: {}!", recordName)
            throw LetsencryptException("No hosted zone found")
        }

        val recordValue = formatChallengeValue(challengeDigest)
        val recordSet = ResourceRecordSet(recordName, RRType.TXT)
                .withTTL(60L)
                .withResourceRecords(ResourceRecord(recordValue))
        val changeBatch = ChangeBatch(ImmutableList.of<Change>(
                Change(action, recordSet)))

        val request = ChangeResourceRecordSetsRequest(hostedZone.get().id, changeBatch)

        /* Commit change and wait until AWS confirms propagation */
        var changeInfo = route53.changeResourceRecordSets(request).changeInfo

        log.info("Applied {} to record {} in zone {} with value {}. Status: {}",
                action, recordName, hostedZone.get().name, recordValue, changeInfo.status)

        while (changeInfo.status == "PENDING") {
            changeInfo = route53.getChange(GetChangeRequest(changeInfo.id)).changeInfo
            Thread.sleep(1000)
        }

        return hostedZone.get()
    }

    /**
     * If no hosted zone is supplied explicitly this function will attempt to find a managed zone in
     * AWS that matches the supplied record.
     * If multiple matching zones are found the most specific one is chosen.
     */
    fun findHostedZone(record: String): Option<HostedZone> {
        val fqdnRecord = determineFqdnRecord(record)
        // Attempt to find the correct hosted zone by matching the longest matching suffix.
        val matchingZone = route53.listHostedZones().hostedZones
                .filter { fqdnRecord.endsWith(it.name) }
                .reduce { acc, zone ->
                    if (zone.name.length > acc.name.length) {
                        zone
                    } else if (zone.name == acc.name) {
                        log.debug("Both zones {} and {} are the same length, " +
                                "checking if private", zone.id, acc.id)
                        // If equal and 'zone' is not private, use that,
                        // else default back to 'acc' (normal behavior).
                        if( !zone.config.isPrivateZone) {
                            log.debug("Zone {} is public, using it", zone.id)
                            zone
                        } else {
                            acc
                        }
                    } else {
                        acc
                    }
                }.toOption()

        matchingZone.forEach { log.info("Found matching zone {} for record {}", it.name, record) }
        return matchingZone
    }

    // Route53 requires TXT record values to be explicitly enclosed in quotation marks.
    private fun formatChallengeValue(challenge: String): String {
        if (challenge.endsWith("\"")) {
            return challenge
        } else {
            return "\"" + challenge + "\""
        }
    }
}